File: node/src/collaboration/contract.ts
/**
 * VerificationContract — first-class type for contract-driven development.
 *
 * Contracts travel in the executor's system partition (never compressed) and
 * are given to the verifier alongside the artifact. The verifier never sees
 * the executor's implementation history — only the goal, contract, and artifact.
 */

export interface AcceptanceCriterion {
  /** Stable id — matched in ContractCheckResult. */
  id: string
  /** Human-readable statement of what correct looks like. */
  text: string
  /** If true, failure here fails the entire contract regardless of score. */
  required: boolean
  /** Contribution to the weighted overall score [0.0–1.0]. */
  weight: number
  /** If true the SDK can verify this deterministically (e.g. word count, schema). */
  machineCheckable: boolean
}

export interface VerificationContract {
  /** Stable id — doubles as the skill name on successful extraction. */
  id: string
  /** Goal this contract governs. Injected into the executor's context. */
  goal: string
  acceptance: AcceptanceCriterion[]
  /** Patterns the executor must avoid. Checked by the verifier. */
  antiPatterns: string[]
  /** Artifacts that must be present before the verifier runs. */
  evidenceRequired: string[]
}

export interface ContractCheckResult {
  criterionId: string
  passed: boolean
  evidence?: string
}

/** Build a contract incrementally. */
export class ContractBuilder {
  private contract: VerificationContract

  constructor(id: string, goal: string) {
    this.contract = { id, goal, acceptance: [], antiPatterns: [], evidenceRequired: [] }
  }

  criterion(
    id: string,
    text: string,
    opts: { required?: boolean; weight?: number; machineCheckable?: boolean } = {},
  ): this {
    this.contract.acceptance.push({
      id,
      text,
      required: opts.required ?? true,
      weight: Math.min(1, Math.max(0, opts.weight ?? 1.0)),
      machineCheckable: opts.machineCheckable ?? false,
    })
    return this
  }

  antiPattern(pattern: string): this {
    this.contract.antiPatterns.push(pattern)
    return this
  }

  evidence(item: string): this {
    this.contract.evidenceRequired.push(item)
    return this
  }

  build(): VerificationContract {
    return structuredClone(this.contract)
  }
}

/**
 * Render the contract as a markdown block for injection into a system prompt.
 * Used by CreatorVerifierBody to inject into the executor's system partition.
 */
export function formatContractForSystemPrompt(contract: VerificationContract): string {
  const lines: string[] = [
    `## Verification Contract: ${contract.id}`,
    "",
    `Goal: ${contract.goal}`,
    "",
    "### Acceptance Criteria",
  ]

  contract.acceptance.forEach((c, i) => {
    const req = c.required ? "[REQUIRED]" : "[OPTIONAL]"
    lines.push(`${i + 1}. ${req} ${c.text} (id: \`${c.id}\`, weight: ${c.weight.toFixed(1)})`)
  })

  if (contract.antiPatterns.length > 0) {
    lines.push("", "### Anti-Patterns (must avoid)")
    contract.antiPatterns.forEach(p => lines.push(`- ${p}`))
  }

  if (contract.evidenceRequired.length > 0) {
    lines.push("", "### Required Evidence")
    contract.evidenceRequired.forEach(e => lines.push(`- ${e}`))
  }

  return lines.join("\n")
}

/** Derive a flat string[] of criterion texts for RuntimeRunner criteria input. */
export function contractToCriteriaStrings(contract: VerificationContract): string[] {
  return contract.acceptance.map(c => c.text)
}

File: python/deepstrike/collaboration/contract.py
"""
VerificationContract — first-class type for contract-driven development.

Contracts travel in the executor's system partition (never compressed) and
are given to the verifier alongside the artifact. The verifier never sees
the executor's implementation history — only the goal, contract, and artifact.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class AcceptanceCriterion:
    """A single verifiable acceptance criterion."""
    id: str
    text: str
    required: bool = True
    weight: float = 1.0
    machine_checkable: bool = False

    def __post_init__(self) -> None:
        self.weight = max(0.0, min(1.0, self.weight))


@dataclass
class VerificationContract:
    """
    First-class contract type: defines what correct looks like before execution starts.

    A VerificationContract is injected into the executor's system partition so it
    survives context renewal and compression. The verifier receives the contract
    alongside the artifact and checks each criterion independently, without access
    to the executor's implementation history.
    """
    id: str
    goal: str
    acceptance: list[AcceptanceCriterion] = field(default_factory=list)
    anti_patterns: list[str] = field(default_factory=list)
    evidence_required: list[str] = field(default_factory=list)


@dataclass
class ContractCheckResult:
    criterion_id: str
    passed: bool
    evidence: Optional[str] = None


class ContractBuilder:
    """Fluent builder for VerificationContract."""

    def __init__(self, id: str, goal: str) -> None:
        self._contract = VerificationContract(id=id, goal=goal)

    def criterion(
        self,
        id: str,
        text: str,
        *,
        required: bool = True,
        weight: float = 1.0,
        machine_checkable: bool = False,
    ) -> "ContractBuilder":
        self._contract.acceptance.append(
            AcceptanceCriterion(
                id=id, text=text, required=required,
                weight=weight, machine_checkable=machine_checkable,
            )
        )
        return self

    def anti_pattern(self, pattern: str) -> "ContractBuilder":
        self._contract.anti_patterns.append(pattern)
        return self

    def evidence(self, item: str) -> "ContractBuilder":
        self._contract.evidence_required.append(item)
        return self

    def build(self) -> VerificationContract:
        import copy
        return copy.deepcopy(self._contract)


def format_contract_for_system_prompt(contract: VerificationContract) -> str:
    """Render the contract as a markdown block for injection into a system prompt."""
    lines = [
        f"## Verification Contract: {contract.id}",
        "",
        f"Goal: {contract.goal}",
        "",
        "### Acceptance Criteria",
    ]
    for i, c in enumerate(contract.acceptance, 1):
        req = "[REQUIRED]" if c.required else "[OPTIONAL]"
        lines.append(f"{i}. {req} {c.text} (id: `{c.id}`, weight: {c.weight:.1f})")

    if contract.anti_patterns:
        lines += ["", "### Anti-Patterns (must avoid)"]
        lines += [f"- {p}" for p in contract.anti_patterns]

    if contract.evidence_required:
        lines += ["", "### Required Evidence"]
        lines += [f"- {e}" for e in contract.evidence_required]

    return "\n".join(lines)


def contract_to_criteria_strings(contract: VerificationContract) -> list[str]:
    """Derive criterion text for kernel in-run gates and compatibility-free body adapters."""
    return [c.text for c in contract.acceptance]
